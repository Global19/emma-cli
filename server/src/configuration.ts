import * as e from 'fp-ts/lib/Either'
import * as t from 'io-ts'

const StarterConfiguration = t.intersection([
  t.type({
    name: t.string,
    path: t.string,
  }),
  t.partial({
    description: t.string,
  }),
])

export interface StarterConfiguration
  extends t.TypeOf<typeof StarterConfiguration> {}

const Configuration = t.type({
  starters: t.array(StarterConfiguration),
})

export interface EmmaConfiguration extends t.TypeOf<typeof Configuration> {}

/**
 * Decodes configurations. Returns error message as a first parameter,
 * or null meaning that configuration is valid.
 *
 * @param config
 */
export function decodeConfiguration(
  config: any,
): e.Either<t.Errors, EmmaConfiguration> {
  return Configuration.decode(config)
}
