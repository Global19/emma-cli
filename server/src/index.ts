import algoliasearch, {
  Index as AlgoliaIndex,
  Task as AlgoliaTask,
} from 'algoliasearch'
import * as e from 'fp-ts/lib/Either'
import * as t from 'io-ts'
import { PathReporter } from 'io-ts/lib/PathReporter'
import probot, { Octokit } from 'probot'

import { Photon, Starter, BatchPayload } from '@generated/photon'

/* Constants */

const EmmaConfigurationFilePath = 'emma.yml'

/* Webhooks */

module.exports = (app: probot.Application) => {
  /* Info */
  app.log.info('Emma server up ðŸš€')

  /* Prisma */

  const photon = new Photon({})

  /* Algolia */

  const algolia = algoliasearch(
    process.env.ALGOLIA_APP_ID!,
    process.env.ALGOLIA_API_KEY!,
  )

  const algoliaStartersIndexName = process.env.ALGOLIA_STARTERS_INDEX!
  const algoliaStartersIndex = algolia.initIndex(algoliaStartersIndexName)

  app.log.info(`Algolia starters index: ${algoliaStartersIndexName}`)

  /* API */

  const api = app.route('/api')

  api.get('/starter/:signiture/', async (req, res) => {
    const starter = await photon.starters.findOne({
      where: { signature: req.body.signature },
    })

    /* Process response */
    if (!starter) {
      res.sendStatus(404)
    } else {
      res.setHeader('Content-Type', 'application/json')
      res.send(
        JSON.stringify({
          ref: starter.ref,
          path: starter.path,
          owner: starter.owner,
          repo: starter.repo,
        }),
      )
    }
  })

  /* Events */

  app.on('push', async context => {
    const owner = context.payload.repository.owner.login
    const repo = context.payload.repository.name
    const ref = context.payload.ref
    const masterRef = `refs/heads/${context.payload.repository.default_branch}`

    /* Ignore changes on non-default ref. */
    if (masterRef !== ref) {
      context.log.info(`Ignoring ref: ${ref}. Default ref: ${masterRef}`)
      return
    }

    /* Load configuration. */
    const configFile = await context.config(EmmaConfigurationFilePath)
    const eConfig = decodeConfiguration(configFile)

    /* Terminate faulty configuration. */
    if (e.isLeft(eConfig)) {
      const report = PathReporter.report(eConfig)
      context.log.debug(
        { report, errors: eConfig.left },
        `Couldn't load configuration.`,
      )
      return
    }

    const config = eConfig.right

    context.log.debug({ config, owner, repo }, 'received configuration')

    try {
      /* Loads starters. */
      const starters = await Promise.all(
        config.starters.map<Promise<EmmaStarter>>(s =>
          loadStarter(context.github, { repo, owner, ref }, s),
        ),
      )

      /* Syncs starters with database. */
      const syncRes = await Promise.all(
        starters.map(starter =>
          saveStarter(photon, algoliaStartersIndex, starter),
        ),
      )

      context.log.debug({ res: syncRes, starters }, `Synced starters.`)

      const cleanRes = await cleanRepositoryStarters(
        photon,
        algoliaStartersIndex,
        { repo, owner },
        starters,
      )

      context.log.debug(
        { repo, owner, res: cleanRes },
        `Cleaned starters in the database.`,
      )
    } catch (err) {
      /* Log error on failure. */
      context.log.warn(err)
    }
  })
}

/* Helper functions. */

/* SCHEMA */

const EmmaStarterConfiguration = t.intersection([
  t.type({
    name: t.string,
    path: t.string,
  }),
  t.partial({
    description: t.string,
  }),
])

interface EmmaStarterConfiguration
  extends t.TypeOf<typeof EmmaStarterConfiguration> {}

const EmmaConfiguration = t.type({
  starters: t.array(EmmaStarterConfiguration),
})

interface EmmaConfiguration extends t.TypeOf<typeof EmmaConfiguration> {}

/**
 * Decodes configurations. Returns error message as a first parameter,
 * or null meaning that configuration is valid.
 *
 * @param config
 */
function decodeConfiguration(
  config: any,
): e.Either<t.Errors, EmmaConfiguration> {
  return EmmaConfiguration.decode(config)
}

/* LOADER */

interface EmmaStarter {
  /* Meta */
  repo: string
  owner: string
  /* Info */
  signature: string
  path: string
  ref: string
  /* Search */
  name: string
  description?: string
  dependencies: string[]
}

/**
 * Loads a starter information from Github.
 *
 * @param github
 * @param param
 * @param config
 */
async function loadStarter(
  github: Octokit,
  { repo, owner, ref }: { repo: string; owner: string; ref: string },
  config: EmmaStarterConfiguration,
): Promise<EmmaStarter> {
  /* Attempt to laod configuration. */
  const res = await github.repos.getContents({
    owner: owner,
    repo: repo,
    ref: ref,
    path: `${config.path}/package.json`,
  })

  switch (res.status) {
    case 200: {
      /* Make sure response is a file. */
      if (Array.isArray(res.data) || !res.data.content) {
        throw new Error(
          `Expected a file at ${config.path} received ${typeof res.data}`,
        )
      }

      const buffer = Buffer.from(res.data.content, 'base64').toString()
      const pkg = JSON.parse(buffer)

      /* Make sure response includes all parameters. */
      if (!pkg.name || !pkg.dependencies) {
        throw new Error(`Missing a starter name or dependencies.`)
      }

      return {
        /* Meta */
        repo: repo,
        owner: owner,
        /* Info */
        signature: `${owner}/${repo}:${config.name}`,
        path: config.path,
        ref: ref,
        /* Search */
        name: config.name,
        description: withDefault(pkg.description, config.description),
        dependencies: Object.keys(pkg.dependencies),
      }
    }
    default: {
      /* Process the error. */
      throw new Error(
        `[${res.status}] Couldn't load package.json at ${config.path}.`,
      )
    }
  }
}

/* PRISMA */

/**
 * Writes a starter to database.
 *
 * @param photon
 * @param starter
 */
async function saveStarter(
  photon: Photon,
  algolia: AlgoliaIndex,
  starter: EmmaStarter,
): Promise<[Starter, AlgoliaTask]> {
  /* Save to database. */
  const prismaRes = await photon.starters.upsert({
    where: { signature: starter.signature },
    create: {
      signature: starter.signature,
      repo: starter.repo,
      owner: starter.owner,
      path: starter.path,
      ref: starter.ref,
      name: starter.name,
      description: starter.description,
      dependencies: { set: starter.dependencies },
    },
    update: {
      repo: starter.repo,
      owner: starter.owner,
      path: starter.path,
      ref: starter.ref,
      name: starter.name,
      description: starter.description,
      dependencies: { set: starter.dependencies },
    },
  })

  const algoliaRes = await algolia.addObject({
    objectID: starter.signature,
    owner: starter.owner,
    repo: starter.repo,
    name: starter.name,
    description: starter.description,
    downloads: 0,
    dependencies: starter.dependencies,
  })

  return [prismaRes, algoliaRes]
}

/**
 * Deletes legacy starters from the database.
 *
 * @param starters
 */
async function cleanRepositoryStarters(
  photon: Photon,
  index: AlgoliaIndex,
  { repo, owner }: { repo: string; owner: string },
  starters: EmmaStarter[],
): Promise<[Starter[], AlgoliaTask, BatchPayload]> {
  const startersNames = starters.map(starter => starter.name)

  const dbStarters = await photon.starters.findMany({
    where: {
      repo: repo,
      owner: owner,
      name: { notIn: startersNames },
    },
  })

  const algoliaRes = await index.deleteObjects(
    dbStarters.map(starter => starter.signature),
  )

  const prismaRes = await photon.starters.deleteMany({
    where: {
      repo: repo,
      owner: owner,
      name: { notIn: startersNames },
    },
  })

  return [dbStarters, algoliaRes, prismaRes]
}

/* UTILS */

/**
 * Uses a default fallback value on undefined or null.
 *
 * @param fallback
 * @param value
 */
function withDefault<T>(fallback: T, value: T | undefined | null): T {
  if (value === null || value === undefined) return fallback
  else return value
}
