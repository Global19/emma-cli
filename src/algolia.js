import algoliasearch from 'algoliasearch'

export const algolia = {
   appId: 'OFCNCOG2CU',
   apiKey: 'f54e21fa3a2a0160595bb058179bfb1e',
   indexName: 'npm-search'
}

const client = algoliasearch(algolia.appId, algolia.apiKey).initIndex(algolia.indexName)

export const search = query => new Promise((resolve, reject) => {
   client.search(query, (err, res) => {
      if (err) {
         reject(err)
      } else {
         resolve(res)
      }
   })
})
