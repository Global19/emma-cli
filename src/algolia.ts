import algoliasearch from 'algoliasearch'

/* Config */

const algolia = {
  appId: 'OFCNCOG2CU',
  apiKey: '6fe4476ee5a1832882e326b506d14126',
  indexName: 'npm-search',
}

const client = algoliasearch(algolia.appId, algolia.apiKey).initIndex(
  algolia.indexName,
)

/* Client */

export interface IPackage {
  name: string
  version: string
  description: string
  owner: string
  humanDownloadsLast30Days: string
}

/**
 *
 * Performs a search for the specified query and returns information
 * displayed in the UI.
 *
 * @param query
 * @param limit
 */
export const getSearch = async (
  query: string,
  limit?: number,
): Promise<IPackage[]> => {
  const { hits } = await client.search<IPackage>({
    query,
    attributesToRetrieve: [
      'name',
      'version',
      'description',
      'owner',
      'humanDownloadsLast30Days',
    ],
    offset: 0,
    length: limit,
  })

  return hits
}
