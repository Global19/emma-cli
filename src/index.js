#!/usr/bin/env node

import meow from 'meow'
import { h, render } from 'ink'
import updateNotifier from 'update-notifier'

import emma from './emma'

// Notify updater
const pkg = require(`../package.json`)

updateNotifier({ pkg }).notify()

// CLI

const cli = meow(
  `
  Usage
    $ emma

  Options
    --dev -D      Add to dev dependencies.
    --limit -L      Number of packages shown, defaults to 5.

  Example
    $ emma -D

  Run without package-name to enter live search.
  Use keyboard to search through package library.
  Use up/down to select packages.
  Use enter to select a package.
  Use tab to move between search/suggestions.
  When query is empty use backspace to remove packages.
  Click space to trigger the install.   
`,
  {
    flags: {
      dev: {
        type: 'boolean',
        alias: 'D',
      },
      limit: {
        type: 'string',
        alias: 'L',
        default: '5',
      },
    },
  },
)

const main = () => {
  let unmount // eslint-disable-line prefer-const

  const onError = () => {
    unmount()
    process.exit(1)
  }

  const onExit = () => {
    unmount()
    process.exit()
  }

  const { dev, limit } = cli.flags

  // Uses `h` instead of JSX to avoid transpiling this file
  unmount = render(h(emma, { dev, limit, onError, onExit }))
}

main()
