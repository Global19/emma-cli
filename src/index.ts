import React from 'react'

export class Emma extends React.Component {
  render() {
    return 'Hey'
  }
}
