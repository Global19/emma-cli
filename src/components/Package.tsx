import React from 'react'
import { Box, Text, StdinContext } from 'ink'

import { IPackage } from '../algolia'
import { WithStdin } from '../utils'

interface Props {
  pkg: IPackage
  active: boolean
  onClick: (pkg: IPackage) => void
}

class Package extends React.Component<WithStdin<Props>> {
  componentDidMount() {
    const { stdin, setRawMode } = this.props

    setRawMode!(true)
    stdin.on('data', this.handleInput)
  }

  componentWillUnmount() {
    const { stdin, setRawMode } = this.props

    stdin.removeListener('data', this.handleInput)
    setRawMode!(false)
  }

  handleInput = (data: any) => {
    const { active, onClick, pkg } = this.props
    const s = String(data)

    if (active && s == 's') {
      onClick(pkg)
    }
  }

  render() {
    const { pkg } = this.props

    return (
      <Box>
        <Box>
          <Text bold>{pkg.name}</Text>
        </Box>
      </Box>
    )
  }
}

export default class PackageWithStdin extends React.Component<Props> {
  render() {
    return (
      <StdinContext.Consumer>
        {({ stdin, setRawMode }) => (
          <Package {...this.props} stdin={stdin} setRawMode={setRawMode} />
        )}
      </StdinContext.Consumer>
    )
  }
}
