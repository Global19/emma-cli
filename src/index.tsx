import React from 'react'
import { Box, StdinContext } from 'ink'

import { IPackage, search } from './algolia'

import { Footer } from './components/Footer'
import Overview from './components/Overview'
import Package from './components/Package'
import Scroll from './components/Scroll'
import Search from './components/Search'

import { IDependency, getNextDependencyType } from './installer'
import { WithStdin } from './utils'

const ARROW_DOWN = '\u001B[B'
const ENTER = '\r'

interface State {
  view: 'SEARCH' | 'SCROLL' | 'OVERVIEW'
  query: string
  hits: IPackage[]
  page: number
  loading: boolean
  dependencies: IDependency[]
}

class Emma extends React.Component<WithStdin<{}>, State> {
  state: State = {
    view: 'SEARCH',
    query: '',
    page: 0,
    hits: [],
    loading: false,
    dependencies: [],
  }

  constructor(props: WithStdin<{}>) {
    super(props)

    this.handleQueryChange = this.handleQueryChange.bind(this)
    this.handleWillReachEnd = this.handleWillReachEnd.bind(this)
  }

  componentDidMount() {
    const { stdin, setRawMode } = this.props

    if (setRawMode) setRawMode(true)
    stdin.on('data', this.handleInput)
  }

  componentWillUnmount() {
    const { stdin, setRawMode } = this.props

    stdin.removeListener('data', this.handleInput)
    if (setRawMode) setRawMode(false)
  }

  handleInput = (data: any) => {
    const s = String(data)

    switch (this.state.view) {
      case 'SEARCH': {
        if (s === ARROW_DOWN || s === ENTER) {
          this.setState({ view: 'SCROLL' })
        }
      }

      case 'SCROLL': {
        if (s === ENTER) {
          this.setState({ view: 'OVERVIEW' })
        }
      }

      case 'OVERVIEW': {
      }
    }
  }

  /**
   * Whenever input changes, switch to the initial screen, change the value
   * of the query accordingly, reset pagination and perform search.
   */
  async handleQueryChange(value: string) {
    this.setState({
      query: value,
      page: 0,
      view: 'SEARCH',
      loading: true,
    })

    const hits = await search(value)

    this.setState({ hits, loading: false })
  }

  /**
   * Start querying new hits and update pagination.
   */
  async handleWillReachEnd() {
    const { query, hits } = this.state
    const page = this.state.page + 1

    const newHits = await search(query, page)

    this.setState({
      page,
      hits: [...hits, ...newHits],
    })
  }

  /**
   * Creates a new dependency if newly selected or toggles the existing one.
   */
  toggleDependency = (pkg: IPackage) => {
    const { dependencies } = this.state
    const dependency = dependencies.find(dep => dep.name === pkg.name)

    if (dependency === undefined) {
      this.setState({
        dependencies: dependencies.concat({
          name: pkg.name,
          type: 'dependency',
        }),
      })
    } else {
      const nextType = getNextDependencyType(dependency)

      if (nextType) {
        this.setState({
          dependencies: dependencies.map(d => {
            if (d.name === pkg.name) return { ...d, type: nextType }
            else return d
          }),
        })
      } else {
        this.setState({
          dependencies: dependencies.filter(d => d.name !== pkg.name),
        })
      }
    }
  }

  render() {
    const { view, query, loading } = this.state

    return (
      <Box flexDirection="column">
        <Search
          value={query}
          onChange={this.handleQueryChange}
          loading={loading}
          active
        />
        <Scroll
          placeholder="Start typing or change query so we can find something!"
          values={this.state.hits}
          onWillReachEnd={this.handleWillReachEnd}
          active={view === 'SCROLL'}
        >
          {pkg => (
            <Package
              key={pkg.objectID}
              pkg={pkg}
              onClick={this.toggleDependency}
              active={pkg.active}
            />
          )}
        </Scroll>
        <Overview />
        <Footer />
      </Box>
    )
  }
}

export default class EmmaWithStdin extends React.Component {
  render() {
    return (
      <StdinContext.Consumer>
        {({ stdin, setRawMode }) => (
          <Emma {...this.props} stdin={stdin} setRawMode={setRawMode} />
        )}
      </StdinContext.Consumer>
    )
  }
}
