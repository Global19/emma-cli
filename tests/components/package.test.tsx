import React from 'react'
import { render } from 'ink-testing-library'

import { IPackage } from '../../src/algolia'
import Package from '../../src/components/Package'

const pkg: IPackage = {
  objectID: 'test-pkg',
  name: 'test-pkg',
  description: 'test-description',
  humanDownloadsLast30Days: '30m',
  version: '2.0.0',
  owner: {
    avatar: 'avatar',
    link: 'link',
    name: 'owner',
    email: 'email',
  },
  repository: {
    url: 'https://github.com/pkg/pkg',
  },
}

describe('package', () => {
  beforeEach(() => {
    jest.restoreAllMocks()
  })

  test('renders package', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={undefined} active={false} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('renders dependency', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'dependency'} active={false} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('renders devDependency', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'devDependency'} active={false} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('renders active package', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={undefined} active={true} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })
  test('renders active depedency', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'dependency'} active={true} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })
  test('renders active devDependency', async () => {
    const fn = jest.fn()
    const { frames, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'devDependency'} active={true} />,
    )

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('shows details', async () => {
    const fn = jest.fn()
    const { frames, stdin, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'devDependency'} active={true} />,
    )

    stdin.write('\u001B[C')

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('hides details', async () => {
    const fn = jest.fn()
    const { frames, stdin, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'devDependency'} active={true} />,
    )

    stdin.write('\u001B[C')
    stdin.write('\u001B[D')

    unmount()
    expect(frames).toMatchSnapshot()
  })

  test('hides details', async () => {
    const mockFn = jest.fn()

    jest.mock('opn', () => ({
      __esModule: true,
      default: mockFn,
    }))

    const fn = jest.fn()
    const { frames, stdin, unmount } = render(
      <Package pkg={pkg} onClick={fn} type={'devDependency'} active={true} />,
    )

    stdin.write('\u001B[C')
    stdin.write('\u001B[C')

    unmount()
    expect(frames).toMatchSnapshot()
    expect(mockFn).toBeCalled()
  })
})
