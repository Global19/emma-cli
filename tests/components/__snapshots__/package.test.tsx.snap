// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package hides details 1`] = `
Array [
  "› 30m test-pkg owner",
  "
› 30m test-pkg owner
  test-description
",
  "› 30m test-pkg owner",
  "› 30m test-pkg owner",
]
`;

exports[`package hides details 2`] = `
Array [
  "› 30m test-pkg owner",
  "
› 30m test-pkg owner
  test-description
",
  "
› 30m test-pkg owner
  test-description
",
  "
› 30m test-pkg owner
  test-description
",
]
`;

exports[`package renders active depedency 1`] = `
Array [
  "› 30m test-pkg owner",
  "› 30m test-pkg owner",
]
`;

exports[`package renders active devDependency 1`] = `
Array [
  "› 30m test-pkg owner",
  "› 30m test-pkg owner",
]
`;

exports[`package renders active package 1`] = `
Array [
  "› 30m test-pkg owner",
  "› 30m test-pkg owner",
]
`;

exports[`package renders dependency 1`] = `
Array [
  "◉ 30m test-pkg owner",
  "◉ 30m test-pkg owner",
]
`;

exports[`package renders devDependency 1`] = `
Array [
  "◉ 30m test-pkg owner",
  "◉ 30m test-pkg owner",
]
`;

exports[`package renders package 1`] = `
Array [
  "  30m test-pkg owner",
  "  30m test-pkg owner",
]
`;

exports[`package shows details 1`] = `
Array [
  "› 30m test-pkg owner",
  "
› 30m test-pkg owner
  test-description
",
  "
› 30m test-pkg owner
  test-description
",
]
`;
